### DISCUSSION TITLE
Caraya PPL Build

### DISCUSSION THREAD

**Discussion by francois-normandin** [JKISoftware/Caraya#128] [Q&A]
Following up on discussions in thread #127 , we'd like to hear your thoughts on how you'd expect to use Caraya when you build it into a PPL. How should it be integrated in the project and LV IDE? What are the use cases for using the PPL (plugins, builds, deployments considerations, palette, quickdrop, etc.)?

Some questions I personally have:

- Would we need Caraya.lvlib and Caraya.lvlibp to both be available from the palettes?
- Should we programmatically change the PPL icon colors so that we do not confuse the LVLIB and PPL VIs on our block diagrams?
- Should the package installation give you a choice of installing either lvlib or lvlibp in vi.lib?
- Should there be a separate package (extension) for building Caraya PPL, instead of supporting this feature in the base package?
- What about deploying in a single project instead of globally? (Dragon-style)

**Comment by jimkring** [comment]
Here are my thoughts on why I would want to use Caraya in PPL form:
- Improved load-time performance of projects that use Caraya (I'm assuming this, but I haven't tested it)
- Better encapsulation of package dependencies, so that use of Caraya doesn't impose limitations on what other packages/VIs I use (this can be achieved without PPLs, too, but PPLs tend to feel safer, in terms of how the encapsulation works)

In terms of how it's implemented, I very rarely need to look inside the Caraya sources, unless I'm working on (debugging?) Caraya and want to see how it works. So, I don't really need a way to switch back and forth, or know whether I'm working with the PPL or the source (don't need different icon colors for PPL or source).  I sort of just expect it to work the same way in PPL form or Source form (meaning the palettes and quickdrop would work the same way and I wouldn't see a difference in how I drop the Caraya support VIs onto my block diagrams).

**Comment by Gerritmerecs** [comment]
Late reply:
We at merecs liked the UI of caraya very much and tried to reuse this for automated tests in test rigs. This is our use case for a PPL. Works very well with only a few modifications.
I have two questions though.
- Many classes are private in caraya. Is this on purpose? I would like to simply overwrite or add a few functions and keep the source prestine as possible. My supervisor now has modified it to tackle a few simple tasks in contrast to the approach Mr. Normandin gave me for my custom report.
- We experience trouble with PPl all the time and the openg packages. Our approach is to explicitly add those to the build description (Build was broken on the LV2020 branch i think). 
![image](https://github.com/user-attachments/assets/29da8fe0-6484-4cc5-9805-c4d95a50ecbb)  Having said this, i would really like to have Caraya deployed as a PPL without having to worry about build descriptions for different LV versions.

Best regards and greetings from germany
Gerrit Feldkamp
